Create events for when send streams become blocked

hawkinsw opened the pull request:
In light of use cases like #261, it seems like a good idea
to create/expose a DataBlocked event (a parallel of a DataWritable
event) when a send stream becomes blocked because of a lack
of credits.

hawkinsw commented:
This is my first substantive patch and I am submitting tentatively. I hope that it is at least worthy of your review. I am eager to see your suggestions for improving it and to make the requested changes. Please let me know if there are others who I should tag for review instead -- I just took github's advice on who to request :-)

Thanks for all the support you've given me to date.
Will

hawkinsw commented:
> This seems like a good change, but it would be easier to tell if there were tests that showed the value of the change.

Funny you mentioned tests! As soon as I submitted this, I thought, "Wow, I should add a test.". Sorry that I didn't include one initially. 

The updated PR I just submitted include a test that, _I think_, addresses your concern. As usual, thank you!

agrover commented:
I'm not clear why we need a new event here. The key async piece seems to be when the stream is unblocked, and there already is an event for that.

Maybe we could tackle this from the other direction -- fixing the issue first mentioned in #261 and seeing what events are needed to work efficiently? Dragana's [comment](https://github.com/mozilla/neqo/issues/261#issuecomment-549727369) could be a good place to start -- either don't add a stream to `have_data_to_send` until unblocked, or maybe add a `blocked` bool to the item in the collection, and only try sending if blocked=false. I'm guessing `Http3Transaction::send()` will likely need to return either a bool (blocked true/false) or usize (bytes sent) instead of `()` so that you can know if lower layers could accept everything or not, and may mean you have to wait to receive a StreamWritable event before trying again.

hawkinsw commented:
> I'm not clear why we need a new event here. The key async piece seems to be when the stream is unblocked, and there already is an event for that.
> 
> Maybe we could tackle this from the other direction -- fixing the issue first mentioned in #261 and 

Thanks for the feedback, @agrover ! A great suggestion to use that approach and one that I was already taking. I simply broke things in two because I thought this was a conceptually different issue and was trying to do the right thing.  

> seeing what events are needed to work efficiently? Dragana's [comment](https://github.com/mozilla/neqo/issues/261#issuecomment-549727369) could be a good place to start -- either don't add a stream to `have_data_to_send` until unblocked, or maybe add a `blocked` bool to the item in 

I agree that this is definitely one part of the optimization. I, too, agree that to do this we do *not* need the event introduced here. As you rightly say, the key async piece is that the stream is writable and we already have an event for that. 

I introduced this event to solve what I thought was the other part of the optimization. Here's the case I was considering:
1. The Http3 server generates a response and inserts it into the underlying transport (with `set_response`). 
1. Only part of that response can be sent before the stream runs out of credit.

In this case, any subsequent invocations of `set_response` can easily tell whether the underlying stream is blocked (as you rightly pointed out) and we could optimize without recourse to this event. However, any part of the Http3 server that continues to attempt to send the outstanding bytes would have no indication that the stream is blocked and, in an act of futility, attempt to resend those as it services the open streams. 

It would be possible for the Http3 server  to set a flag on its reference to the stream if/when `set_response` said that the stream is blocked (if we modified that function in the way you suggested below). However, that would assume that there would be subsequent calls to `set_response`. The Http3 server could also poll the stream during its event loop to check for the status and set the proper flag then. That is certainly an option and would not require this event.

I am sure that I am not thinking through the issue the way you are, as someone who has great experience with the codebase and the protocol. However, I wanted to write this so you didn't think that I just slapped something together and submitted it for review. :-)

Thank again you for the response and I hope this made some sort of sense!

Will

> the collection, and only try sending if blocked=false. I'm guessing `Http3Transaction::send()` will likely need to return either a bool (blocked true/false) or usize (bytes sent) instead of `()` so that you can know if lower layers could accept everything or not, and may mean you have to wait to receive a StreamWritable event before trying again.

hawkinsw commented:
Rebased the PR to be compatible with master after @ddragana's refactor of Http3Connection.

hawkinsw commented:
@agrover and @dragana, I was wondering if you could take another look at this and provide me some guidance? Thanks!

agrover commented:
It's hard for me to assess fully because this PR doesn't include the changes that would make use of the new event, only the addition of the event itself. Would it be possible to also see another commit that uses the events to implement the http3 processing improvement you describe?

hawkinsw commented:
Absolutely! As soon as I resolve the documentation issue I'll return to
this and post a follow up!

Thanks for the note!

On Thu, Jan 9, 2020, 8:41 PM Andy Grover <notifications@github.com> wrote:

> It's hard for me to assess fully because this PR doesn't include the
> changes that would make use of the new event, only the addition of the
> event itself. Would it be possible to also see another commit that uses the
> events to implement the http3 processing improvement you describe?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/neqo/pull/345?email_source=notifications&email_token=ACCP2CVOZ7H7JF573U7ULVDQ47G6BA5CNFSM4JXZNKWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEISMMWI#issuecomment-572835417>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCP2CWZBFNA5ZX2BHTVVJDQ47G6BANCNFSM4JXZNKWA>
> .
>

agrover commented:
If this PR is going to be reviewed and move forward I think it's probably better to open a fresh one. Closing.
